core/plugin_api.rs

```diff
@@ -11,13 +11,31 @@
 pub use crate::Buf;
 pub use crate::Op;
 pub use crate::OpId;
+pub use crate::Resource;
+pub use crate::ResourceId;
+pub use crate::ResourceTable;
 pub use crate::ZeroCopyBuf;
 
 pub type InitFn = fn(&mut dyn Interface);
 
 pub type DispatchOpFn =
   fn(&mut dyn Interface, &[u8], Option<ZeroCopyBuf>) -> Op;
 
+/// Equivilent to ResourceTable for normal ops, but only includes
+/// non generic("boxed") versions of get, get_mut, and remove.
+/// We can't have generic functions on a trait and use it as a object
+/// this is a limtation of rust.
+pub trait WrappedResourceTable {
+  fn has(&self, rid: ResourceId) -> bool;
+  fn get_boxed(&self, rid: ResourceId) -> Option<&dyn Resource>;
+  fn get_mut_boxed(&mut self, rid: ResourceId) -> Option<&mut dyn Resource>;
+  fn add(&mut self, name: &str, resource: Box<dyn Resource>) -> ResourceId;
+  fn entries(&self) -> Vec<(ResourceId, String)>;
```

Comment:
Let's not give plugins access to that.

Reply:
It might be helpful for testing. Any reason not to?

Reply:
> It might be helpful for testing. Any reason not to?

It's going to be a public API. The reason not to is that we only want to add things to the public API that we're willing to maintain forever.

We can consider this feature when someone brings up an important use case that really needs it. For testing we can have internal APis.

Also, if anything, it should be an iterator.